### PR TITLE
Update `cli-template-contracts-hardhat`

### LINKED ISSUE #646: Update `cli-template-contracts-hardhat`

### PR DESCRIPTION
**Pull request by vplasencia**
## Related Issue(s)

Closes #646

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] My changes generate no new warnings
-   [x] I have run `yarn prettier` and `yarn lint` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by 0xjei — APPROVED** [review]
Great work! :fire: 

I noticed the following [error](https://hardhat.org/hardhat-runner/docs/errors#HH308) when running `yarn dev`. This should be related to the `localhost` parameter. Let me approve the PR but could you please check before merging? Thank you! 

image

**Comment by vplasencia** [comment]
Hey! @0xjei thanks for the review. Yes, the project is prepared to use `npm` instead of `yarn`. To use `yarn`, you need to change some configs first. The other two templates are using `yarn` now, but I'm changing the config to use `npm` by default in a way that it's easy to use other package managers too with minimal changes. 

The idea is to have `npm` by default in all the templates and add info in the docs on how devs can easily add other package managers like `yarn` and `pnpm`.

**Comment by 0xjei** [comment]
> Hey! @0xjei thanks for the review. Yes, the project is prepared to use `npm` instead of `yarn`. To use `yarn`, you need to change some configs first. The other two templates are using `yarn` now, but I'm changing the config to use `npm` by default in a way that it's easy to use other package managers too with minimal changes.
> 
> The idea is to have `npm` by default in all the templates and add info in the docs on how devs can easily add other package managers like `yarn` and `pnpm`.

Do you think that really makes sense? It could be confusing as the whole project uses `yarn` as the default!

**Comment by 0xjei** [comment]
> > Hey! @0xjei thanks for the review. Yes, the project is prepared to use `npm` instead of `yarn`. To use `yarn`, you need to change some configs first. The other two templates are using `yarn` now, but I'm changing the config to use `npm` by default in a way that it's easy to use other package managers too with minimal changes.
> > The idea is to have `npm` by default in all the templates and add info in the docs on how devs can easily add other package managers like `yarn` and `pnpm`.
> 
> Do you think that really makes sense? It could be confusing as the whole project uses `yarn` as the default!

BTW, I'm still getting the same error with `npm run dev` too

**Comment by vplasencia** [comment]
> Do you think that really makes sense? It could be confusing as the whole project uses yarn as the default!

The idea is to create the templates as generic as possible in a way that dev can use the package manager they prefer.

What do you think that would be better?

File: packages/cli-template-contracts-hardhat/.eslintignore [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+# debug
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+
+# packages
+cli-template-*
+
+snark-artifacts
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.eslintignore]: This folder shouldn't be part of this template anymore. So shouldn't be necessary to have it here

File: packages/cli-template-contracts-hardhat/.eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    "root": true,
+    "env": {
+        "es6": true
+    },
+    "extends": ["airbnb-base", "airbnb-typescript/base", "prettier"],
+    "parser": "@typescript-eslint/parser",
+    "parserOptions": {
+        "ecmaVersion": 6,
+        "sourceType": "module",
+        "project": ["./tsconfig.json", "./packages/**/tsconfig.json"]
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.eslintrc.json]: This isn't a monorepo. `"./packages/**/tsconfig.json"` is not necessary

**Review by 0xjei — APPROVED** [review]
Works fine now and makes sense to use `yarn` V4 by default as we internally discussed. Thank you! :blush:

File: packages/cli-template-contracts-hardhat/.eslintignore [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+.DS_Store
+*.pem
+
+# debug
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+
+# packages
+cli-template-*
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.eslintignore]: This is not necessary here

File: packages/cli-template-contracts-hardhat/.eslintignore [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+# types
+types
+
+# circuits
+circuits
+
+# production
+dist
+build
+docs
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.eslintignore]: These folders shouldn't be necessary here

File: packages/cli-template-contracts-hardhat/.prettierignore [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+coverage
+coverage.json
+
+# hardhat
+artifacts
+cache
+typechain-types
+packages/contracts/deployed-contracts/undefined.json
+packages/contracts/deployed-contracts/hardhat.json
+packages/contracts/deployed-contracts/localhost.json
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.prettierignore]: Shouldn't be necessary

File: packages/cli-template-contracts-hardhat/.prettierignore [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+# contracts
+Verifier*.sol
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.prettierignore]: There's no such a contract here

File: packages/cli-template-contracts-hardhat/.prettierignore [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+# misc
+.DS_Store
+*.pem
+
+# debug
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+
+zk-artifacts
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/.prettierignore]: No necessary

File: packages/cli-template-contracts-hardhat/README.md [hunk 8]
```diff
@@ -4,13 +4,19 @@ This project demonstrates a basic Semaphore use case. It comes with a sample con
 
 ## Usage
 
-### Compile
+### Install dependencies
+
+```bash
+yarn
+```
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/README.md:11]: This is usually another `## Install` section in the other packages, right before the `## Usage` one.

e.g.

```md
## Install

Install the dependencies:

`` `bash
yarn
`` `

## Usage

...
```

File: packages/cli-template-contracts-hardhat/package.json [hunk 9]
```diff
@@ -9,57 +9,67 @@
...
+        "prettier-plugin-solidity": "^1.3.1",
+        "solhint": "^4.1.1",
         "solidity-coverage": "^0.8.1",
         "ts-node": "^10.9.2",
         "typechain": "^8.3.0",
         "typescript": "^5.3.3"
     },
     "dependencies": {
-        "@semaphore-protocol/contracts": "4.0.0-alpha.8"
+        "@semaphore-protocol/contracts": "^4.0.0-alpha.8"
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/package.json]: It should be good practice to not use ^ here, as it's a dependency

File: packages/cli-template-contracts-hardhat/test/Feedback.ts [hunk 10]
```diff
@@ -43,28 +47,38 @@ describe("Feedback", () => {
...
+
+            for (const user of users) {
+                await feedbackContract.joinGroup(user.commitment)
+                group.addMember(user.commitment)
+            }
+
             const feedback = encodeBytes32String("Hello World")
 
-            const proof = await generateProof(users[1], group, feedback, groupId)
+            const fullProof = await generateProof(users[1], group, feedback, groupId)
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/test/Feedback.ts]: We can call this `proof` now, as there's no `proof.proof` anymore (it's `proof.points` now).

File: packages/cli-template-contracts-hardhat/tsconfig.json [hunk 11]
```diff
@@ -1,15 +1,11 @@
...
-        "module": "CommonJS",
-        "strict": true,
+        "target": "es2020",
+        "module": "commonjs",
         "esModuleInterop": true,
-        "outDir": "dist",
-        "typeRoots": ["node_modules/@types", "types"]
-    },
-    "include": ["scripts/**/*", "tasks/**/*", "test/**/*", "build/typechain/**/*", "types/**/*"],
-    "files": ["./hardhat.config.ts"]
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/tsconfig.json]: Not sure if it includes these folders by default

File: packages/cli-template-contracts-hardhat/package.json [hunk 12]
```diff
@@ -9,57 +9,67 @@
...
+        ".editorconfig",
         ".env.example",
+        ".eslintignore",
+        ".eslintrc.json",
+        ".prettierignore",
+        ".prettierrc.json",
         "hardhat.config.ts",
-        "tsconfig.json",
-        "README.md"
+        "tsconfig.json"
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/package.json]: This list is missing: `README.md`, `.yarn/`, `.yarnrc.yml`

It doesn't need: `scripts/`

File: packages/cli-template-contracts-hardhat/package.json [hunk 13]
```diff
@@ -9,57 +9,67 @@
...
+        "dev": "hardhat node & yarn deploy --network localhost",
         "compile": "hardhat compile",
-        "deploy": "yarn compile && hardhat deploy",
+        "deploy": "hardhat compile && hardhat deploy",
+        "verify": "hardhat verify",
         "test": "hardhat test",
         "test:report-gas": "REPORT_GAS=true hardhat test",
         "test:coverage": "hardhat coverage",
         "typechain": "hardhat typechain",
-        "prepublish": "tar -czf files.tgz .gitignore"
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/package.json:28]: We need this

**Review by cedoor — CHANGES_REQUESTED** [review]

File: packages/cli-template-contracts-hardhat/tsconfig.json [hunk 14]
```diff
@@ -1,14 +1,12 @@
...
+        "module": "commonjs",
         "esModuleInterop": true,
-        "outDir": "dist",
-        "typeRoots": ["node_modules/@types", "types"]
+        "forceConsistentCasingInFileNames": true,
+        "strict": true,
+        "skipLibCheck": true,
+        "resolveJsonModule": true
     },
     "include": ["scripts/**/*", "tasks/**/*", "test/**/*", "build/typechain/**/*", "types/**/*"],
```
Comment (cedoor) [inline packages/cli-template-contracts-hardhat/tsconfig.json]:
```suggestion
    "include": ["scripts/**/*", "tasks/**/*", "test/**/*", "typechain-types/**/*"],
```

The same change needs to be applied to the other templates.

**Review by cedoor — CHANGES_REQUESTED** [review]

**Review by cedoor — APPROVED** [review]

**Pull request merged**